File: polymorphism/polym.go
package main

import (
	"fmt"
)

type Forma interface {
	Area() int64
	Perimeter() float64
}

type Comportamento interface {
	Falar()
	Rodar()
}

type Quadrado struct {
	Lado float64
}

func (q Quadrado) Area() int64 {
	return int64(q.Lado * q.Lado)
}

func (c Quadrado) Perimeter() float64 {
	return 4 * c.Lado
}

type Circulo struct {
	Raio float64
}

func (c Circulo) Area() int64 {
	return int64(3.1416 * c.Raio * c.Raio)
}

func (c Circulo) Perimeter() float64 {
	return 2 * 3.1416 * c.Raio
}

func (c Circulo) Falar() {
	fmt.Println("ola sou um circulo")
}

func (c Circulo) Rodar() {
	fmt.Println("rodando")
}

// Função que usa polimorfismo: aceita qualquer "Forma"
func imprimirInfo(f Forma) {
	fmt.Println("Área:", f.Area())
	fmt.Println("Perimeter:", f.Perimeter())
}

func imprimirInfo2(f Comportamento) {
	f.Falar()
	f.Rodar()
}

func main() {
	quadrado := Quadrado{Lado: 4}
	circulo := Circulo{Raio: 5}

	imprimirInfo(quadrado)
	fmt.Println("----------------")
	imprimirInfo(circulo)
	imprimirInfo2(circulo)
}

File: polymorphism/polym.java
// Definimos una interfaz "Forma"
interface Forma {
    double area();
    double perimeter();
}

// Implementamos la interfaz en la clase "Cuadrado"
class Cuadrado implements Forma {
    private double lado;

    public Cuadrado(double lado) {
        this.lado = lado;
    }

    @Override
    public double area() {
        return lado * lado;
    }

    @Override
    public double perimeter() {
        return 4 * lado;
    }
}

// Implementamos la interfaz en la clase "Círculo"
class Circulo implements Forma {
    private double radio;

    public Circulo(double radio) {
        this.radio = radio;
    }

    @Override
    public double area() {
        return Math.PI * radio * radio;
    }

    @Override
    public double perimeter() {
        return 2 * Math.PI * radio;
    }
}

// Clase principal
public class PolimorfismoEjemplo {
    public static void imprimirArea(Forma forma) {
        System.out.println("Área: " + forma.area());
    }

    public static void main(String[] args) {
        Forma cuadrado = new Cuadrado(4);
        Forma circulo = new Circulo(5);

        imprimirArea(cuadrado);
        imprimirArea(circulo);
    }
}
